bpe.go

```diff
@@ -287,3 +309,202 @@ func (m Model) DecodeFromStream(reader io.Reader) ([]string, error) {
 	}
 	return sentences, nil
 }
+
+type encodingToken struct {
+	id   TokenID
+	prev int
+	next int
+}
+
+type mergeEvent struct {
+	priority int
+	pos      int
+}
+
+type mergeQueue []*mergeEvent
+
+func (mq mergeQueue) Len() int { return len(mq) }
+
+func (mq mergeQueue) Less(i, j int) bool {
+	return mq[i].priority < mq[j].priority ||
+		mq[i].priority == mq[j].priority && mq[i].pos < mq[j].pos
+}
+
+func (mq mergeQueue) Swap(i, j int) {
+	mq[i], mq[j] = mq[j], mq[i]
+}
+
+func (mq *mergeQueue) Push(x interface{}) {
+	*mq = append(*mq, x.(*mergeEvent))
+}
+
+func (mq *mergeQueue) Pop() interface{} {
+	old := *mq
+	n := len(old)
+	item := old[n-1]
+	old[n-1] = nil // avoid memory leak
+	*mq = old[0 : n-1]
+	return item
+}
+
+// EncodeSentence accepts string of space-separated words and tokenizes each word
+// according to the BPE rules. Through encodingConfig one can state whether to add BOS, EOS tokens
+// and whether to reverse the output sequences. EncodeSentence returns both numerical encoding
+// and sequence of text tokens of the sentence.
+func (m Model) EncodeSentence(sentence string, encodingConfig EncodingConfig,
+) (EncodedString, []string, error) {
```

Comment:
It's only possible to do it backwards - from string to id, because of unknown tokens. In yttm they leave the unknown tokens during tokenization and not replace them with `<UNK>` (so if we know only `a`, `b` tokens, in yttm `abcrt` -> `a b crt`, not `abcrt` -> `a b <UNK>`). 
That's why it's impossible to infer tokens from strings encoding (I wanted to do that in the beginning).

Now I think about it, do we really want it to be like that too?

Reply:
Got it. I can imagine no use-case which depends on the unknown strings after splitting. Let's drop `[]string` completely then?

Reply:
The guys state the thing as the `tokenization tool`, guess that's why they do it. But usually when they use BPE there are `<UNK>`s as the goal is to have a fixed-vocabulary-size splits.

Anyway, for our use-case we don't need strings, so I guess it's fine to drop it completely. If we need it some time, it will be easy to get back, I will leave the commits untouched:)

Reply:
Sounds good.